docs/guides/scaling_crawlers.mdx

```diff
@@ -0,0 +1,58 @@
+
+
+As we build our crawler, we might want to control how many tasks we do to the website at a time. Crawlee provides several options to fine tune how many parallel tasks should be made at any time, how many tasks should be done per minute, and how should scaling work based on the available system resources.
+
+:::tip
+
+All of these options are available on all crawlers Crawlee provides, but for this guide we'll be using the <ApiLink to="class/BeautifulSoupCrawler">`BeautifulSoupCrawler`</ApiLink>. We can see all the options for concurrency settings <ApiLink to="class/ConcurrencySettings">`here`</ApiLink>.
+:::
+
+## `max_tasks_per_minute`
+
+This controls how many total tasks can be processed per minute. It counts the amount of tasks done every second, to ensure there is not a burst of tasks at the `max_concurrency` limit followed by a long period of waiting. By default, it is set to `Infinity` which means the crawler will keep going up to the `max_concurrency`. We would set this if we wanted our crawler to work at full throughput, but also not keep hitting the website we're crawling with non-stop requests.
+
+<CodeBlock language="python">
+    {MaxRequestsPerMinuteSource}
+</CodeBlock>
+
+## `min_concurrency` and `max_concurrency`
+
+These control how many parallel tasks can be run at any time. By default, crawlers will start with one parallel task at a time and scale up over time to a maximum of 200 requests at a time.
+
+:::caution Don't set `min_concurrency` too high!
+
+Setting this option too high compared to the available system resources will make your crawler run extremely slow or might even crash.
+
+It's recommended to leave it at the default value that is provided and letting the crawler scale up and down automatically based on available resources instead.
+
+:::
+
+## `desired_concurrency`
+
+This option specifies the amount of tasks that should be running in parallel at the start of the crawler, assuming there are so many available. It defaults to the same value as `min_concurrency`.
+
+
+<CodeBlock language="python">
+    {MinMaxConcurrencySource}
+</CodeBlock>
+
+## `AutoscaledPool`
+
+`AutoscaledPool` manages a pool of asynchronous resource-intensive tasks that are executed in parallel. The pool only starts new tasks if there is enough free CPU and memory available. If an exception is thrown in any of the tasks, it is propagated and the pool is stopped.
+
+Internally each crawler has an `AutoscaledPool` which takes the above `ConcurrencySettings` as parameter to fine tune the operation. Some of the properties it has are-
+- `desired_concurrency_ratio` - The minimum ratio of concurrency to reach before more scaling up is allowed (a number between `0` and `1`). By default, it is set to `0.9`.
+- `scale_up_step_ratio` and `scale_down_step_ratio` - These values define the fractional amount of desired concurrency to be added or subtracted as the autoscaling pool scales up or down. Both of these values default to `0.05`.
+
```

Comment:
```suggestion
---
id: scaling-crawlers
title: Scaling crawlers
description: Learn how to scale your crawlers by controlling concurrency and limiting requests per minute.
---

import ApiLink from '@site/src/components/ApiLink';

import CodeBlock from '@theme/CodeBlock';

import MaxRequestsPerMinuteSource from '!!raw-loader!./code/max_tasks_per_minute.py';
import MinMaxConcurrencySource from '!!raw-loader!./code/min_max_concurrency.py';

As we build our crawler, we may want to control how many tasks it performs at any given time. In other words, how many requests it makes to the web we are trying to scrape. Crawlee offers several options to fine-tune the number of parallel tasks, limit the number of requests per minute, and optimize scaling based on available system resources.

:::tip

All of these options are available across all crawlers provided by Crawlee. In this guide, we are using the <ApiLink to="class/BeautifulSoupCrawler">`BeautifulSoupCrawler`</ApiLink> as an example. You should also explore the <ApiLink to="class/ConcurrencySettings">`ConcurrencySettings`</ApiLink>.

:::

## Max tasks per minute

The `max_tasks_per_minute` setting in <ApiLink to="class/ConcurrencySettings">`ConcurrencySettings`</ApiLink> controls how many total tasks the crawler can process per minute. It ensures that tasks are spread evenly throughout the minute, preventing a sudden burst at the `max_concurrency` limit followed by idle time. By default, this is set to `Infinity`, meaning the crawler can run at full speed, limited only by `max_concurrency`. Use this option if you want to throttle your crawler to avoid overwhelming the target website with continuous requests.

<CodeBlock className="language-python">
    {MaxRequestsPerMinuteSource}
</CodeBlock>

## Minimum and maximum concurrency

The `min_concurrency` and `max_concurrency` options in the <ApiLink to="class/ConcurrencySettings">`ConcurrencySettings`</ApiLink> define the minimum and maximum number of parallel tasks that can run at any given time. By default, crawlers start with a single parallel task and gradually scale up to a maximum of concurrent requests.

:::caution Avoid setting minimum concurrency too high

If you set `min_concurrency` too high compared to the available system resources, the crawler may run very slowly or even crash. It is recommended to stick with the default value and let the crawler automatically adjust concurrency based on the system's available resources.

:::

## Desired concurrency

The `desired_concurrency` option in the <ApiLink to="class/ConcurrencySettings">`ConcurrencySettings`</ApiLink> specifies the initial number of parallel tasks to start with, assuming sufficient resources are available. It defaults to the same value as `min_concurrency`.

<CodeBlock className="language-python">
    {MinMaxConcurrencySource}
</CodeBlock>

## Autoscaled pool

The <ApiLink to="class/AutoscaledPool">`AutoscaledPool`</ApiLink> manages a pool of asynchronous, resource-intensive tasks that run in parallel. It automatically starts new tasks only when there is enough free CPU and memory. To monitor system resources, it leverages the <ApiLink to="class/Snapshotter">`Snapshotter`</ApiLink> and <ApiLink to="class/SystemStatus">`SystemStatus`</ApiLink> classes. If any task raises an exception, the error is propagated, and the pool is stopped. Every crawler uses an <ApiLink to="class/AutoscaledPool">`AutoscaledPool`</ApiLink> under the hood.
```